pkg/agent/sysadvisor/plugin/qosaware/resource/cpu/region/provisionpolicy/policy_rama.go

```diff
@@ -0,0 +1,146 @@
+}
+
+func NewPolicyRama(regionName string, regionType types.QoSRegionType, ownerPoolName string,
+	conf *config.Configuration, extraConfig interface{}, metaReader metacache.MetaReader,
+	metaServer *metaserver.MetaServer, emitter metrics.MetricEmitter) ProvisionPolicy {
+	p := &PolicyRama{
+		PolicyBase:                 NewPolicyBase(regionName, regionType, ownerPoolName, metaReader, metaServer, emitter),
+		configuredIndicatorMetrics: []string{},
+		controllers:                make(map[string]helper.PIDController),
+	}
+
+	// initialize indicator metrics region interested in
+	indicatorMetrics, ok := conf.PolicyRama.IndicatorMetrics[regionType]
+	if !ok {
+		klog.Warningf("[qosaware-cpu-rama] indicator metrics not found for region %v", regionType)
+		return p
+	}
+	p.configuredIndicatorMetrics = indicatorMetrics
+
+	// initialize pid controllers for every indicator of this region
+	for _, metricName := range p.configuredIndicatorMetrics {
+		params, ok := conf.PolicyRama.PIDParameters[metricName]
+		if !ok {
+			klog.Warningf("[qosaware-cpu-rama] pid parameter not found for indicator %v", metricName)
+			continue
+		}
+		p.controllers[metricName] = helper.NewPIDController(params)
+	}
+
+	return p
+}
+
+func (p *PolicyRama) Update() error {
+	// sanity check
+	if err := p.sanityCheck(); err != nil {
+		return err
+	}
+
+	cpuSize := p.ControlKnobs[types.ControlKnobNonReclaimedCPUSize].Value
+
+	// pass the current value to regulator to avoid slow start from zero after restart
+	p.regulator.SetLatestCPURequirement(int(cpuSize))
+
+	cpuSizeAdjustedGlobal := math.Inf(-1)
+
+	// run pid control for each indicator
+	for _, metricName := range p.configuredIndicatorMetrics {
+		indicator := p.Indicators[metricName]
+		controller := p.controllers[metricName]
+
+		cpuSizeAdjusted := controller.Adjust(cpuSize, indicator.Target, indicator.Current)
+		if cpuSizeAdjusted > cpuSizeAdjustedGlobal {
+			cpuSizeAdjustedGlobal = cpuSizeAdjusted
+		}
+	}
+	p.regulator.Regulate(cpuSizeAdjustedGlobal)
+
+	return nil
+}
+
+func (p *PolicyRama) sanityCheck() error {
+	var (
+		isLegal bool
+		errList []error
+	)
+
+	// 1. check control knob legality
+	isLegal = true
+	if p.ControlKnobs == nil || len(p.ControlKnobs) <= 0 {
+		isLegal = false
+	} else {
+		v, ok := p.ControlKnobs[types.ControlKnobNonReclaimedCPUSize]
+		if !ok || v.Value <= 0 {
+			isLegal = false
+		}
+	}
+	if !isLegal {
+		errList = append(errList, fmt.Errorf("illegal control knob %v", p.ControlKnobs))
+	}
+
+	// 2. check indicator legality
+	isLegal = true
+	for _, metricName := range p.configuredIndicatorMetrics {
+		indicator, ok := p.Indicators[metricName]
+		if !ok {
+			isLegal = false
+		} else if indicator.Target <= 0 || indicator.Current <= 0 {
+			isLegal = false
+		}
+	}
+	if !isLegal {
+		errList = append(errList, fmt.Errorf("illegal indicator %v", p.Indicators))
+	}
+
+	// 3. check pid controllers
+	for _, metricName := range p.configuredIndicatorMetrics {
```

Comment:
will it be better to return error during new functions? since this configuredIndicatorMetrics will always return error here, and will never works as expected anymore

Reply:
PID controller pamameters may be dynamically changed through KCC in the future so we need dynamic check here.